### PR TITLE
mm: fix realloc overwrite memcpy when use a new block

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary
mm: fix realloc overwrite memcpy when use a new block

## Impact
mm

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
This patch need cherry-pick to 10.2 to avoid the buffer overflow error.

File: mm/mm_heap/mm_realloc.c [hunk 1]
```diff
@@ -355,7 +355,7 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
       newmem = (FAR void *)mm_malloc(heap, size);
       if (newmem)
         {
-          memcpy(newmem, oldmem, oldsize);
+          memcpy(newmem, oldmem, oldsize - SIZEOF_MM_ALLOCNODE);
```
Comment (Ouss4) [inline mm/mm_heap/mm_realloc.c:358]: was the following `free` failing here?

Comment (xiaoxiang781216) [reply mm/mm_heap/mm_realloc.c:358]: No, the problem isn't fee, but memcpy. oldsize include the house keeping data(SIZEOF_MM_ALLOCNODE) before oldmem, memcpy with oldsize will overwrite SIZEOF_MM_ALLOCNODE bytes of the next region.

Comment (Ouss4) [reply mm/mm_heap/mm_realloc.c:358]: Ah I see, I thought we are messing wit the memory that's going to be freed, it's actually the other pointer.

**Review by Ouss4 — APPROVED** [review]
LGTM.

**Comment by xiaoxiang781216** [comment]
BTW, this bug is actually found by KASAN: https://github.com/apache/incubator-nuttx/pull/4743.

**Pull request merged**